Mifare.c

```diff
@@ -1,203 +1,406 @@
+            CHECK_STATUS(status);
+
+            /* Configure Discovery loop for Poll Mode */
+            status = phacDiscLoop_SetConfig(&sDiscLoop, PHAC_DISCLOOP_CONFIG_NEXT_POLL_STATE, PHAC_DISCLOOP_POLL_STATE_DETECTION);
+            CHECK_STATUS(status);
+
+            /* Run Discovery loop */
+            status = phacDiscLoop_Run(&sDiscLoop, PHAC_DISCLOOP_ENTRY_POINT_POLL);
+
+        }while((status & PH_ERR_MASK) != PHAC_DISCLOOP_DEVICE_ACTIVATED); /* Exit on Card detection */
+
+        /* Card detected */
+        /* Get the tag types detected info */
+        status = phacDiscLoop_GetConfig(&sDiscLoop, PHAC_DISCLOOP_CONFIG_TECH_DETECTED, &wTagsDetected);
+
+        /* Check for Status */
+        if ((status & PH_ERR_MASK) == PH_ERR_SUCCESS)
+        {
+            /* Check for Type A tag detection */
+            if (PHAC_DISCLOOP_CHECK_ANDMASK(wTagsDetected, PHAC_DISCLOOP_POS_BIT_MASK_A))
+            {
+
+							uint8_t byteBufferSize = sDiscLoop.sTypeATargetInfo.aTypeA_I3P3[0].bUidSize + 1;
+					    uint8_t byteBuffer[sDiscLoop.sTypeATargetInfo.aTypeA_I3P3[0].bUidSize + 1];
+						        uint8_t i;
+					        char asciiBuffer[sDiscLoop.sTypeATargetInfo.aTypeA_I3P3[0].bUidSize + 1];
+					
+					        if (byteBufferSize + 1 > sDiscLoop.sTypeATargetInfo.aTypeA_I3P3[0].bUidSize + 1) {
+					            // Truncate if we got back too much data
+					            byteBufferSize = sDiscLoop.sTypeATargetInfo.aTypeA_I3P3[0].bUidSize;
+					        }
+					
+					        for (i = 0; i < byteBufferSize; i++) {
+					            sprintf(&asciiBuffer[2 * i], "%02X", sDiscLoop.sTypeATargetInfo.aTypeA_I3P3[0].aUid[i]);
+					        }
+					
+					        return PyUnicode_FromString(asciiBuffer);
+    
+
+            }
         }
-
-        return PyUnicode_FromString(asciiBuffer);
     }
 
     Py_RETURN_NONE;
 }
 
 PyObject *Mifare_read_block(Mifare *self, PyObject *args)
 {
-    const size_t bufferSize = PHAL_MFUL_READ_BLOCK_LENGTH;
-    uint8_t data[bufferSize];
-    uint8_t blockIdx;
-    phStatus_t status;
-
-    if (!PyArg_ParseTuple(args, "b", &blockIdx)) {
+   	    uint8_t blockIdx;
+	if (!PyArg_ParseTuple(args, "b", &blockIdx)) {
         return NULL;
     }
 
-    status = phalMful_Read(&self->data.alMful, blockIdx, data);
 
-    if (status != PH_ERR_SUCCESS) {
-        return PyErr_Format(ReadError, "Read failed: %04x", status);
-    }
+
+
+    
+    
+phStatus_t  status = 0;
+    uint16_t    wTagsDetected = 0;
+    uint8_t     bUid[PHAC_DISCLOOP_I3P3A_MAX_UID_LENGTH];
+    uint8_t     bUidSize;
+                       
+
+                        status = phalMful_Read(&salMfc, blockIdx, bDataBuffer);
+                            if (status != PH_ERR_SUCCESS) {
+						        return PyErr_Format(ReadError, "Read failed: %04x", status);
+						    }
 
 #if PY_MAJOR_VERSION >= 3
-    return Py_BuildValue("y#", data, bufferSize);
+                        return Py_BuildValue("y#", &bDataBuffer[0], MFC_BLOCK_DATA_SIZE);
```

Comment:
I believe the write operation only allows writing one page (4 bytes) at a time.  The read operation seems to be returning 16 bytes though.  This is where my knowledge of NFC/RFID is exceedingly thin.  Are all Mifare cards organized in pages of 4 bytes each?  If so, both reads and writes should be limited to 4 bytes since they're being done on a block (page) basis.

I welcome additional insight here.  If there's a different approach that makes more logical sense, I'm all for it.